src/createAction.ts

```diff
@@ -7,33 +7,50 @@ import { Action } from 'redux'
  * @template P The type of the action's payload.
  * @template T the type used for the action type.
  */
-export interface PayloadAction<P = any, T extends string = string>
-  extends Action<T> {
+export type PayloadAction<
+  P = any,
+  T extends string = string,
+  M = void
+> = Action<T> & {
   payload: P
-}
+} & ([M] extends [void] ? {} : { meta: M })
+
+export type Diff<T, U> = T extends U ? never : T
 
-export type Diff<T, U> = T extends U ? never : T;
+export type PrepareAction<P> =
+  | ((...args: any[]) => { payload: P })
+  | ((...args: any[]) => { payload: P; meta: any })
 
 /**
  * An action creator that produces actions with a `payload` attribute.
  */
-export type PayloadActionCreator<P = any, T extends string = string> = { type: T } & (
-  /*
-  * The `P` generic is wrapped with a single-element tuple to prevent the
-  * conditional from being checked distributively, thus preserving unions
-  * of contra-variant types.
-  */
-  [undefined] extends [P] ? {
-    (payload?: undefined): PayloadAction<undefined, T>
-    <PT extends Diff<P, undefined>>(payload?: PT): PayloadAction<PT, T>
-  }
-  : [void] extends [P] ? {
-    (): PayloadAction<undefined, T>
-  }
-  : {
-    <PT extends P>(payload: PT): PayloadAction<PT, T>
-  }
-);
+export type PayloadActionCreator<
+  P = any,
+  T extends string = string,
+  PA extends PrepareAction<P> | void = void
+> = {
+  type: T
+} & (PA extends (...args: any[]) => any
+  ? (ReturnType<PA> extends { meta: infer M }
```

Comment:
Just so you know, my eyes are still bleeding here :)

Reply:
Yeah, that one is wild. But it defines many different behaviours :D

Reply:
FWIW here's how I understand this code. 

PayloadActionCreator always has a `type` property of _type_ `T`

If PA is a function we check if it has a `meta` field. If yes the final type is 
```
{
  type: T extends string,
  meta: M,
  payload: P
}
  ```

If the return value does not have a `meta` field it's
```
{
  type: T extends string,
  payload: P
}
  ```

Now the else, if PA is `not` a function we default back to the older behavior which terrifies me.

@phryneas as some practical feedback, it might be helpful to make it clear to the reader
1. where the if/else statements are
2. what the final type will be in each scenario.

Reply:
Almost, but no.

All Action creators share the `{ type: T }`, but the rest of the functionality differs.

If a `PA` argument is passed, the first branch will be taken.

Within that branch, if PA returns something with `{ meta }`, again take the first branch. Resulting type is a function `(...args: Parameters<PA>) => PayloadAction<P, T, M>` (with the additional `type` from above -  all that follow will, too). Otherwise it's a function `(...args: Parameters<PA>) => PayloadAction<P, T>)`.

Now, on to the other branch - no PA given. This has to handle some special cases first and then falls back to the most used case.

If P is undefined, it returns a method that might take 0 or 1 argument. If P is void, it's a method with 0 arguments. Otherwise it is a method with 1 argument. All these methods infer their argument and return that inferred argument as the payload of the returned action.

Formatting this is in a language that knows only ternaries and in a file that will be shuffled around with prettier is unfortunately something I did not achieve :/